### PR TITLE
FTR-9-Service reservation

### PR DESCRIPTION
**Pull request by mpuskadij**
Major changes:
### Back button component
I made a new component called BackButton:

~~~
@Composable
fun BackButton(onBackPressed : () ->Unit, modifier: Modifier = Modifier) {
IconButton(
        onClick = onBackPressed,
        modifier= modifier
    ) {
        Icon(
            imageVector = Icons.AutoMirrored.Filled.ArrowBack,
            contentDescription = "Back",
            tint = Color.White
        )
    }
}
~~~
Modifier is optional, but the function is not.

### Calendar Component
~~~
fun ReservationCalendar(selectedDates: SelectableDates, onDismiss: () -> Unit, onConfirm : (Long?) -> Unit) {
    val datePickerState = rememberDatePickerState(
        initialSelectedDateMillis = System.currentTimeMillis(),
        selectableDates = selectedDates
    )
    DatePickerDialog(
        onDismissRequest = onDismiss,
        confirmButton = {
            TextButton(onClick = { onConfirm(datePickerState.selectedDateMillis) }) {
                Text(text = stringResource(R.string.confirm))
            }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) {
                Text(text = stringResource(R.string.cancel))
            }
        }
    ) {
        DatePicker(state = datePickerState)
    }
}
~~~
3 parameters:

1. onConfirm - executes when a date is selected, returns to the function selected date in Long, UTC timezone
2.  onDismiss - executes when user clicks outside the date picker or cliks on button "Dismiss"
3. selectableDate - object of class that implements interface SelectableDate which dictates what year and dates can be selected before displaying the date picker

### Switching between CompanyDetails and Service Reservation
Added a new route:
~~~
composable("company/{companyName}/{serviceName}") {
        backStackEntry ->
        val companyName = backStackEntry.arguments?.getString("companyName") ?: "Unknown"
        val serviceName = backStackEntry.arguments?.getString("companyName") ?: "Unknown"
        ServiceReservationScreen(serviceName, companyName, navController)

    }
~~~
Added switching to the new route: in CompanyDetailsContent:
~~~
ProvidedServicesListItem(serviceName =  companyCategory.value, onServiceClicked = {
                        navController.navigate("company/${companyName}/${companyCategory.value}")
                    })
~~~

### CODE REVIEW

**Comment by dpetek21** [comment]
Looks good, well done, master of the craft ⛏️

**Review by dpetek21 — APPROVED** [review]
Looks good, I have faith.

File: app/src/main/java/hr/foi/air/servicesync/navigation/AppNavHost.kt [hunk 1]
```diff
@@ -57,6 +56,13 @@ fun NavGraphBuilder.AppNavHost(navController: NavHostController) {
             companyName = companyName
         )
     }
+    composable("company/{companyName}/{serviceName}") {
```
Comment (ibeusan20) [inline app/src/main/java/hr/foi/air/servicesync/navigation/AppNavHost.kt:61]: Nice. Continuing usage of proper implementation of navigation.

File: app/src/main/java/hr/foi/air/servicesync/ui/components/BackButton.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import androidx.compose.material3.Icon
+import androidx.compose.material3.IconButton
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.unit.dp
+
+@Composable
+fun BackButton(onBackPressed : () ->Unit, modifier: Modifier = Modifier) {
```
Comment (ibeusan20) [inline app/src/main/java/hr/foi/air/servicesync/ui/components/BackButton.kt:15]: I must say this is very well done in terms of reusability of the code. Thanks!

File: app/src/main/java/hr/foi/air/servicesync/ui/components/ReservationCalendar.kt [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import androidx.compose.material3.TextButton
+import androidx.compose.material3.rememberDatePickerState
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.res.stringResource
+import hr.foi.air.servicesync.R
+import hr.foi.air.servicesync.business.PresentAndFutureSelectableDates
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun ReservationCalendar(selectedDates: SelectableDates, onDismiss: () -> Unit, onConfirm : (Long?) -> Unit) {
```
Comment (ibeusan20) [inline app/src/main/java/hr/foi/air/servicesync/ui/components/ReservationCalendar.kt:16]: Made calendar in a way it's compatible for future use. Although it wasn't in the issue, I must thank you for this.

File: app/src/main/java/hr/foi/air/servicesync/ui/components/TestReservationScreen.kt
Comment (ibeusan20) [inline app/src/main/java/hr/foi/air/servicesync/ui/components/TestReservationScreen.kt:1]: Nicely noticed test file for deletion.

File: app/src/main/java/hr/foi/air/servicesync/ui/contents/CompanyDetailsContent.kt [hunk 4]
```diff
@@ -122,7 +122,9 @@ fun CompanyDetailsContent(
                     )
                 },
                 supportingContent = {
-                    ProvidedServicesListItem(companyCategory.value)
+                    ProvidedServicesListItem(serviceName =  companyCategory.value, onServiceClicked = {
+                        navController.navigate("company/${companyName}/${companyCategory.value}")
```
Comment (ibeusan20) [inline app/src/main/java/hr/foi/air/servicesync/ui/contents/CompanyDetailsContent.kt:139]: Nice way of using the new navigation.

File: app/src/main/java/hr/foi/air/servicesync/ui/screens/ServiceReservationScreen.kt [hunk 5]
```diff
@@ -32,13 +33,15 @@ import hr.foi.air.servicesync.backend.FirestoreService
...
+import hr.foi.air.servicesync.ui.components.BackButton
+import hr.foi.air.servicesync.ui.components.ReservationCalendar
 import hr.foi.air.servicesync.ui.components.isDark
 import java.text.DateFormat
 import java.util.*
 
 @OptIn(ExperimentalMaterial3Api::class)
 @Composable
-fun ServiceReservationScreen(serviceName: String, companyId: String) {
+fun ServiceReservationScreen(serviceName: String, companyId: String, navController: NavController) {
```
Comment (ibeusan20) [inline app/src/main/java/hr/foi/air/servicesync/ui/screens/ServiceReservationScreen.kt:44]: Adding navcontroller to servicereservation screen is nicely done

File: app/src/main/java/hr/foi/air/servicesync/ui/screens/ServiceReservationScreen.kt [hunk 6]
```diff
@@ -48,9 +51,11 @@ fun ServiceReservationScreen(serviceName: String, companyId: String) {
     var loading by remember { mutableStateOf(false) }
 
     Column(modifier = Modifier.padding(16.dp)) {
+        BackButton(onBackPressed = { navController.popBackStack()})
         Text(
             text = serviceName,
             style = MaterialTheme.typography.headlineMedium,
+            color = isDark(surfaceVariantLight, surfaceVariantDark),
```
Comment (ibeusan20) [inline app/src/main/java/hr/foi/air/servicesync/ui/screens/ServiceReservationScreen.kt:58]: Nice! Using the isDark the right way! Not nice, it's VERY nice!

File: app/src/main/java/hr/foi/air/servicesync/ui/screens/ServiceReservationScreen.kt [hunk 7]
```diff
@@ -70,41 +75,30 @@ fun ServiceReservationScreen(serviceName: String, companyId: String) {
...
-                                reservationManager,
-                                companyId,
-                                it
-                            ) { slots -> availableSlots = slots }
-                        }
-                        showDatePicker = false
-                    }) {
-                        Text(text = stringResource(R.string.confirm))
-                    }
+            ReservationCalendar(
```
Comment (ibeusan20) [inline app/src/main/java/hr/foi/air/servicesync/ui/screens/ServiceReservationScreen.kt:78]: Revamped use of calendar picker to a reusable one. Amazing work. Reusability is very important!

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -17,4 +17,5 @@
     <string name="confirm">Potvrdi</string>
     <string name="cancel">Odustani</string>
     <string name="select_date">Odaberite datum...</string>
+    <string name="available_appointments">Slobodni termini:</string>
```
Comment (ibeusan20) [inline app/src/main/res/values/strings.xml:21]: True example of good use of strings. Great for future translation!

**Review by ibeusan20 — APPROVED** [review]
Everything is done by the cookbook. I must mention that. It's very well done.

File: app/src/main/java/hr/foi/air/servicesync/navigation/AppNavHost.kt (hunk 1 above)
Comment (dblazevic21) [reply app/src/main/java/hr/foi/air/servicesync/navigation/AppNavHost.kt:61]: Really good usage of nested navigation!

File: app/src/main/java/hr/foi/air/servicesync/ui/components/CompanyNameAndImage.kt [hunk 9]
```diff
@@ -52,18 +53,7 @@ fun CompanyImage(companyName: String, imageUrl: String?, onBackPressed: () -> Un
...
-                .padding(16.dp)
-                .align(Alignment.TopStart)
-        ) {
-            Icon(
-                imageVector = Icons.Filled.ArrowBack,
-                contentDescription = "Back",
-                tint = Color.White
-            )
-        }
+        BackButton(onBackPressed = onBackPressed, modifier =  Modifier.padding(16.dp).align(Alignment.TopStart))
```
Comment (dblazevic21) [inline app/src/main/java/hr/foi/air/servicesync/ui/components/CompanyNameAndImage.kt:56]: Good refactor for the Iconbutton, but button should change color depending on the mode. (The button is white, on light mode and white on dark mode)

**Review by dblazevic21 — APPROVED** [review]
Really well made, but consider using different color approach in the future.

File: app/src/main/java/hr/foi/air/servicesync/ui/components/CompanyNameAndImage.kt (hunk 9 above)
Comment (mpuskadij) [reply app/src/main/java/hr/foi/air/servicesync/ui/components/CompanyNameAndImage.kt:56]: Ok, will fix.

File: app/src/main/java/hr/foi/air/servicesync/ui/components/BackButton.kt (hunk 2 above)
Comment (mpuskadij) [reply app/src/main/java/hr/foi/air/servicesync/ui/components/BackButton.kt:15]: No problem!

File: app/src/main/java/hr/foi/air/servicesync/navigation/AppNavHost.kt (hunk 1 above)
Comment (mpuskadij) [reply app/src/main/java/hr/foi/air/servicesync/navigation/AppNavHost.kt:61]: Thanks!

File: app/src/main/java/hr/foi/air/servicesync/ui/components/ReservationCalendar.kt (hunk 3 above)
Comment (mpuskadij) [reply app/src/main/java/hr/foi/air/servicesync/ui/components/ReservationCalendar.kt:16]: Glad to do it.

**Comment by ibeusan20** [comment]
@mmarkoovin21 please approve this PR so it can be merged, conflicts have been resolved successfully

File: app/src/main/java/hr/foi/air/servicesync/ui/components/ReservationCalendar.kt [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+import androidx.compose.material3.DatePickerDialog
+import androidx.compose.material3.ExperimentalMaterial3Api
+import androidx.compose.material3.SelectableDates
+import androidx.compose.material3.Text
+import androidx.compose.material3.TextButton
+import androidx.compose.material3.rememberDatePickerState
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.res.stringResource
+import hr.foi.air.servicesync.R
+import hr.foi.air.servicesync.business.PresentAndFutureSelectableDates
```
Comment (mmarkoovin21) [inline app/src/main/java/hr/foi/air/servicesync/ui/components/ReservationCalendar.kt]: Remove imports that you don't use.

File: app/src/main/java/hr/foi/air/servicesync/ui/components/BackButton.kt
Comment (mmarkoovin21) [inline app/src/main/java/hr/foi/air/servicesync/ui/components/BackButton.kt:1]: Sorry for the hate but I don't think this is much needed from UX perspective if we have backbutton of on our phones at the bottom. Why should I bring my thumb to the top left corner if bottom right corner is closer.

**Review by mmarkoovin21 — CHANGES_REQUESTED** [review]
Looks really good. Few notes:
- Would like to se **CompanyDetails** screen after booking a service
- Don't think that back button on top of the **CompanyDetails** screen is needed form UX perspective.

File: app/src/main/java/hr/foi/air/servicesync/ui/components/BackButton.kt
Comment (mpuskadij) [reply app/src/main/java/hr/foi/air/servicesync/ui/components/BackButton.kt:1]: Altough the user will use maybe android back, not all android phones have the back and having a screen without a back button may confuse the users that they can't return

File: app/src/main/java/hr/foi/air/servicesync/ui/components/ReservationCalendar.kt (hunk 10 above)
Comment (mpuskadij) [reply app/src/main/java/hr/foi/air/servicesync/ui/components/ReservationCalendar.kt]: Removed the unused import as reported

**Comment by mpuskadij** [comment]
@mmarkoovin21 removed the unused import. As discussed with other team members, it makes more sense the have the back button, so it was not removed. Please review.

**Review by mmarkoovin21 — APPROVED** [review]
Noted!

**Pull request merged**